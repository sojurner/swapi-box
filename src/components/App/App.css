@import url('https://fonts.googleapis.com/css?family=Cinzel');

.App {
  margin: 0;
  position: absolute;
  width: 100vw;
}

.App__TITLE {
  border: 0.2rem solid #9c816b;
  border-radius: 0.4rem;
  margin: 2rem auto;
  padding: 1rem;
  width: 16.5rem;
}

.App__TITLE h1 {
  color: #9c816b;
  font-size: 2.5rem;
  font-family: 'Cinzel', 'STJEDISE', serif;
  text-align: center;
  margin: 0;
}

.fa-galactic-republic,
.fa-chevron-circle-left {
  color: #c9c0b5;
  position: absolute;
  font-size: 3rem;
  top: 2rem;
  cursor: pointer;
  border-radius: 50%;
  border: #c9c0b5;
  padding: 0 0 0 2px;
  height: 3rem;
  width: 3rem;
}

.fa-chevron-circle-left {
  font-size: 1.7rem;
  left: 14rem;
  top: 1.5rem;

  z-index: 100;
  animation: slideInLeft 1.5s cubic-bezier(0.075, 0.82, 0.165, 1);
}

.fa-chevron-circle-left:hover {
  color: rgb(220, 220, 195);
}

.fa-galactic-republic {
  left: 2rem;
}

.fa-galactic-republic:hover {
  color: beige;
  animation: dark-pulse 1.4s linear infinite;
}

.header {
  color: #c9c0b5;
  background: rgba(100, 100, 100, 0.623);
  letter-spacing: 3px;
  padding: 0.8rem 0;
  margin: 1rem auto;
  min-width: 5rem;
  width: 100%;
  max-width: 32rem;
  text-align: center;
  border-radius: 0.7rem;
}

.header h1,
.header h4 {
  margin: 0.2rem;
}

.planets {
  background: url(../../images/BattleFront.jpg) no-repeat center center fixed;
  background-size: cover;
}

.people {
  background: url(../../images/sw-people.jpg) no-repeat center center fixed;
  background-size: cover;
}

.vehicles {
  background: url(../../images/star-wars-droid.jpg) no-repeat center center
    fixed;
  background-size: cover;
}

@keyframes dark-pulse {
  0% {
    box-shadow: 0 0 0 0 rgba(252, 250, 250, 0);
  }
  50% {
    box-shadow: 0 0 0 10px rgba(255, 255, 255, 0.146);
  }

  100% {
    box-shadow: 0 0 0 0 rgba(145, 145, 145, 0.132);
  }
}
