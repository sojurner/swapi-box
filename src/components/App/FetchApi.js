import axios from "axios";

export const FetchApi = async url => {
  const response = await axios.get(`https://swapi.co/api/${url}/`);
  console.log(response);
  const { results, next } = response.data;
  const data = [results];
  const completeList = await FetchNextPage(next, data);
  return completeList;
};

const FetchNextPage = async (url, array) => {
  let nextUrl = url;
  const response = await axios.get(url);
  const { next, results } = response.data;
  const data = [...array, results];
  if (next) {
    nextUrl = next;
    FetchNextPage(nextUrl, data);
  }
  return data;
};

export const mapThroughArray = data => {
  let mapped = data.forEach();
  const x = [...data];
  // return data.results.map(person => {
  //   return {
  //     name: person.name,
  //     homeworld: person.homeworld.name,
  //     species: person.species[0].name,
  //     population: person.species[0].homeworld.population
  //   };
  // });
  console.log(x);
};
