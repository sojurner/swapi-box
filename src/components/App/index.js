import React, { Component } from "react";
import NavBar from "../Navigation/NavBar";
import ContentRoute from "../Navigation/ContentRoute";
import "./App.css";
import { FetchApi, mapThroughArray } from "../Fetch/FetchApi";
import BackgroundScroll from "../BackgroundScroll";

class App extends Component {
  constructor() {
    super();
    this.state = {
      films: null,
      backgroundFilm: null,
      planets: null,
      people: null,
      vehicles: null,
      favorites: null,
      errorStatus: ""
    };
  }

  async componentDidMount() {
    const backgroundScroll = { target: { textContent: "films" } };
    await this.getData(backgroundScroll);
    this.setRandomFilm();
  }

  setRandomFilm = () => {
    const { films } = this.state
    const randomFilmIndex = () => (Math.random()*films.length)+0.5<<0;
    const backgroundFilm = films[randomFilmIndex()];
    this.setState({ backgroundFilm });
  }

  // componentDidMount() {
  //   const backgroundScroll = { target: { textContent: "films" } };
  //   this.getData(backgroundScroll);
  // }

  getData = async event => {
    try {
      let selectedData = event.target.textContent;
      const result = await FetchApi(selectedData);
      // const scrapeData = mapThroughArray(result);
      this.setState({ [selectedData]: result });
    } catch (error) {
      console.log(error);
    }
  };

  markAsFavorite = () => {
    console.log("favorited");
  };

  render() {
    const { people, planets, vehicles, favorites, backgroundFilm } = this.state;
  
    return (
      <div className="App">
        <h1 className="App__TITLE">SWAPI BOX</h1>
       {backgroundFilm && <BackgroundScroll {...backgroundFilm} />}
        <NavBar getData={this.getData} />
        <ContentRoute
          favorites={favorites}
          people={people}
          planets={planets}
          vehicles={vehicles}
        />
      </div>
    );
  }
}

export default App;