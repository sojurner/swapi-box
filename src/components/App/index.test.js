import React from "react";

import App from "./index.js";
import { mockPeople } from "../../MockData/MockData";

describe("App", () => {
  let wrapper;
  let mockEvent;

  beforeEach(async () => {
    localStorage.clear();
    wrapper = shallow(<App />);
    await wrapper.instance().componentDidMount;
  });

  it("should match snapshot with default states", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should match snapshot with background scroll in state ", () => {
    const backgroundFilm = {dont: 'stop'};
    wrapper.setState({backgroundFilm});
    expect(wrapper).toMatchSnapshot();
  });

  describe("getData", () => {
    beforeEach(() => {
      wrapper = shallow(<App />);
      mockEvent = { target: { name: "people" } };
      window.fetch = jest.fn().mockImplementation(() =>
        Promise.resolve({
          json: () => Promise.resolve(mockPeople)
        })
      );
    });

    it("should call FetchApi", async () => {
      await wrapper.instance().getData(mockEvent, 1);
      expect(fetch).toHaveBeenCalled();
    });

    it("should set correct state with the value passed", async () => {
      await wrapper.instance().getData(mockEvent, 1);

      expect(wrapper.state("people").length).toEqual(10);
    });

    it("should set correct error state if value returns failed fetch", async () => {
      window.fetch = jest
        .fn()
        .mockImplementation(() => Promise.reject(new Error("failed")));
      await wrapper.instance().getData(mockEvent);
      expect(wrapper.state("errorStatus")).toEqual("failed");
    });

    it("should call setButtonPressed", () => {
      wrapper.instance().setButtonPressed = jest.fn();
      wrapper.instance().getData(mockEvent);
      expect(wrapper.instance().setButtonPressed).toHaveBeenCalled();
    });

    it("should run fetch only if state is empty or not null", async () => {
      const people = [{}, {}];
      wrapper.setState({ people });
      expect(wrapper.state("people").length).toEqual(2);
      await wrapper.instance().getData(mockEvent);
      expect(wrapper.state("people").length).toEqual(10);
    });

    it("should return if alreadyHasData is true and not a new page", async () => {
      const people = [{}, {}];
      const expected = people;
      await wrapper.setState({ people });
      await wrapper.instance().getData(mockEvent, 1);
      expect(wrapper.state("people")).toEqual(expected);
    });

    it("should not return if alreadyHasData is true and a new page exists", async () => {
      const people = [{}, {}];
      wrapper.setState({ people });
      await wrapper.instance().getData(mockEvent, 2);
      expect(wrapper.state("people").length).toEqual(10);
    });

    describe('setRandomFilm', () => {
      it("should set state backgroundFilm of single film from the films array in state", () => {
        wrapper = shallow(<App />);
        const films = [{}, {}, {}];
        wrapper.setState({ films });
        wrapper.instance().setRandomFilm();
        expect(wrapper.state('backgroundFilm')).toEqual({});
      });

      it("should return if films array in state is null", () => {
        wrapper = shallow(<App />);
        const films = null;
        wrapper.setState({ films });
        wrapper.instance().setRandomFilm();
        expect(wrapper.state('backgroundFilm')).toEqual(null);
      });
    });

    describe('setButtonPressed', () => {
      it("should take a string and set activeButton state", () => {
        wrapper = shallow(<App />);
        wrapper.instance().setButtonPressed('people');
        expect(wrapper.state('activeButton')).toEqual('people');
      });

      it("should revert pageCounter to 1", () => {
        wrapper = shallow(<App />);
        wrapper.setState({ pageCounter: 2 });
        wrapper.instance().setButtonPressed('people');
        expect(wrapper.state('pageCounter')).toEqual(1);
      });
    });

    describe('handleFavorites and toggleFavorites and setFavoritesFromStorage', () => {
      let favorites;
      beforeEach(() => {
        wrapper = shallow(<App />);
        favorites = [{name: 'one', favorite: true}, {name: 'two', favorite: true}, {name: 'three', favorite: true}];
      });

      it("toggleFavorites should toggle card's favorite property ", () => {
        const mockCard = {favorite: false};
        wrapper.instance().toggleFavorites(mockCard);
        expect(mockCard.favorite).toEqual(true);
      });

      it("handleFavorites should add a card to favorites array in state ", () => {
        const mockCard = {name: 'four', favorite: true};
        wrapper.setState({ favorites });
        wrapper.instance().handleFavorites(mockCard);
        expect(wrapper.state('favorites').length).toEqual(4);
      });

      it("handleFavorites should remove a card to favorites array in state ", () => {
        const mockCard = {name: 'three', favorite: false};
        wrapper.setState({ favorites });
        wrapper.instance().handleFavorites(mockCard);
        expect(wrapper.state('favorites').length).toEqual(2);
      });

      it("handleFavorites should update localStorage with a new favorites array ", () => {
        const mockCard = {name: 'four', favorite: true};
        wrapper.setState({ favorites });
        wrapper.instance().handleFavorites(mockCard);
        const expected = JSON.parse(localStorage.getItem('favorites')).length;
        expect(expected).toEqual(4);
      });

      it("setFavoritesFromStorage retreive favorites from localStorage and set to state", () => {
        localStorage.setItem('favorites', JSON.stringify(favorites));
        wrapper.instance().setFavoritesFromStorage();
        let expected = JSON.parse(localStorage.getItem('favorites')).length;
        expect(expected).toEqual(3);
      });

      it("setFavoritesFromStorage should return if favorites from localStorage does not exist", () => {
        favorites = [{nada:'NOTHING HERE'}];
        wrapper.setState({ favorites });
        wrapper.instance().setFavoritesFromStorage();
        expect(wrapper.state('favorites')).toEqual([{nada:'NOTHING HERE'}]);
      });
    });

    describe('handlePage', () => {
      it("should increment state's pageCount by one", () => {
        wrapper = shallow(<App />);
        wrapper.instance().handlePage(true);
        expect(wrapper.state('pageCounter')).toEqual(2);
        wrapper.instance().handlePage(true);
        expect(wrapper.state('pageCounter')).toEqual(3);
      });

      it("should decrement state's pageCount by one if pageCounter not the first page", () => {
        wrapper = shallow(<App />);

        wrapper.instance().handlePage(true);
        wrapper.instance().handlePage(true);
        wrapper.instance().handlePage(true);
        wrapper.instance().handlePage(false);
        expect(wrapper.state('pageCounter')).toEqual(3);
      });

      it("should become empty string if dercremented on first page", () => {
        wrapper = shallow(<App />);

        wrapper.instance().handlePage(false);
        expect(wrapper.state('pageCounter')).toEqual('');  
      });

      it("should set pageCounter to 2 if incrementing on empty string or 0", () => {
        wrapper = shallow(<App />);

        wrapper.instance().handlePage(false);
        expect(wrapper.state('pageCounter')).toEqual('');  

        wrapper.instance().handlePage(true);
        expect(wrapper.state('pageCounter')).toEqual(2);
          
        wrapper.instance().handlePage(false);
        expect(wrapper.state('pageCounter')).toEqual('');  
        wrapper.instance().handlePage(false);
        expect(wrapper.state('pageCounter')).toEqual('');  

        wrapper.instance().handlePage(true);
        expect(wrapper.state('pageCounter')).toEqual(2); 
      });

      it("should default in switch case if no other cases are met", () => {
        wrapper = shallow(<App />);
        wrapper.instance().handlePage('true');
        expect(wrapper.state('pageCounter')).toEqual(1);
        wrapper.instance().handlePage('true');
        expect(wrapper.state('pageCounter')).toEqual(1);
      });
        
    });
  });
});
/* global shallow */