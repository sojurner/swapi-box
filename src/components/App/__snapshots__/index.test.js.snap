// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot with background scroll in state  1`] = `
<div
  className="films App"
>
  <header
    className="App__TITLE"
  >
    <h1>
      SWAPI BOX
    </h1>
  </header>
  <Component
    dont="stop"
  />
  <Component
    activeButton="films"
    favorites={Array []}
    getData={[Function]}
    pressed="films"
  />
  <Component
    favorites={Array []}
    handlePage={[Function]}
    people={null}
    planets={null}
    toggleFavorites={[Function]}
    vehicles={null}
  />
</div>
`;

exports[`App should match snapshot with default states 1`] = `
<div
  className="films App"
>
  <header
    className="App__TITLE"
  >
    <h1>
      SWAPI BOX
    </h1>
  </header>
  <Component
    activeButton="films"
    favorites={Array []}
    getData={[Function]}
    pressed="films"
  />
  <Component
    favorites={Array []}
    handlePage={[Function]}
    people={null}
    planets={null}
    toggleFavorites={[Function]}
    vehicles={null}
  />
</div>
`;
